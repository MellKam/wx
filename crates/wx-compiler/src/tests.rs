use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
