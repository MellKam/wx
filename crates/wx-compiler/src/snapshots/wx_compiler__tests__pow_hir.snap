---
source: crates/wx-compiler/src/tests.rs
expression: hir
---
hir:
  file_id: 0
  functions:
    - name:
        symbol: 1
        span:
          start: 12
          end: 15
      ty:
        params:
          - Primitive: I32
          - Primitive: I32
        result:
          Primitive: I32
      stack:
        scopes:
          - kind: Block
            label: ~
            parent: ~
            locals:
              - name:
                  symbol: 2
                  span:
                    start: 16
                    end: 20
                ty:
                  Primitive: I32
                mutability: ~
                accesses:
                  - span:
                      start: 221
                      end: 225
                    kind: Read
              - name:
                  symbol: 4
                  span:
                    start: 31
                    end: 34
                ty:
                  Primitive: I32
                mutability:
                  start: 27
                  end: 30
                accesses:
                  - span:
                      start: 55
                      end: 58
                    kind: Read
                  - span:
                      start: 149
                      end: 152
                    kind: Read
                  - span:
                      start: 235
                      end: 238
                    kind: ReadWrite
              - name:
                  symbol: 5
                  span:
                    start: 110
                    end: 116
                ty:
                  Primitive: I32
                mutability:
                  start: 106
                  end: 109
                accesses:
                  - span:
                      start: 174
                      end: 180
                    kind: Read
                  - span:
                      start: 211
                      end: 217
                    kind: ReadWrite
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
          - kind: Block
            label: ~
            parent: 0
            locals: []
            inferred_type: Never
            expected_type: ~
          - kind: Loop
            label: ~
            parent: 0
            locals: []
            inferred_type:
              Primitive: I32
            expected_type:
              Primitive: I32
          - kind: Block
            label: ~
            parent: 2
            locals: []
            inferred_type: Never
            expected_type: ~
      block:
        kind:
          Block:
            scope_index: 0
            expressions:
              - kind:
                  IfElse:
                    condition:
                      kind:
                        Binary:
                          operator:
                            kind: Less
                            span:
                              start: 59
                              end: 60
                          left:
                            kind:
                              Local:
                                scope_index: 0
                                local_index: 1
                            span:
                              start: 55
                              end: 58
                            ty:
                              Primitive: I32
                          right:
                            kind:
                              Int: 0
                            span:
                              start: 61
                              end: 62
                            ty:
                              Primitive: I32
                      span:
                        start: 55
                        end: 62
                      ty: Bool
                    then_block:
                      kind:
                        Block:
                          scope_index: 1
                          expressions:
                            - kind:
                                Return:
                                  value:
                                    kind:
                                      Int: 0
                                    span:
                                      start: 80
                                      end: 81
                                    ty:
                                      Primitive: I32
                              span:
                                start: 73
                                end: 81
                              ty: Never
                          result: ~
                      span:
                        start: 63
                        end: 89
                      ty: Never
                    else_block: ~
                span:
                  start: 52
                  end: 89
                ty: Unit
              - kind:
                  LocalDeclaration:
                    name:
                      symbol: 5
                      span:
                        start: 110
                        end: 116
                    scope_index: 0
                    local_index: 2
                    expr:
                      kind:
                        Int: 1
                      span:
                        start: 124
                        end: 125
                      ty:
                        Primitive: I32
                span:
                  start: 100
                  end: 125
                ty: Unit
            result:
              kind:
                Loop:
                  scope_index: 2
                  block:
                    kind:
                      Block:
                        scope_index: 2
                        expressions:
                          - kind:
                              IfElse:
                                condition:
                                  kind:
                                    Binary:
                                      operator:
                                        kind: Eq
                                        span:
                                          start: 153
                                          end: 155
                                      left:
                                        kind:
                                          Local:
                                            scope_index: 0
                                            local_index: 1
                                        span:
                                          start: 149
                                          end: 152
                                        ty:
                                          Primitive: I32
                                      right:
                                        kind:
                                          Int: 0
                                        span:
                                          start: 156
                                          end: 157
                                        ty:
                                          Primitive: I32
                                  span:
                                    start: 149
                                    end: 157
                                  ty: Bool
                                then_block:
                                  kind:
                                    Block:
                                      scope_index: 3
                                      expressions:
                                        - kind:
                                            Break:
                                              scope_index: 2
                                              value:
                                                kind:
                                                  Local:
                                                    scope_index: 0
                                                    local_index: 2
                                                span:
                                                  start: 174
                                                  end: 180
                                                ty:
                                                  Primitive: I32
                                          span:
                                            start: 168
                                            end: 180
                                          ty: Never
                                      result: ~
                                  span:
                                    start: 158
                                    end: 192
                                  ty: Never
                                else_block: ~
                            span:
                              start: 146
                              end: 192
                            ty: Unit
                          - kind:
                              Binary:
                                operator:
                                  kind: MulAssign
                                  span:
                                    start: 218
                                    end: 220
                                left:
                                  kind:
                                    Local:
                                      scope_index: 0
                                      local_index: 2
                                  span:
                                    start: 211
                                    end: 217
                                  ty:
                                    Primitive: I32
                                right:
                                  kind:
                                    Local:
                                      scope_index: 0
                                      local_index: 0
                                  span:
                                    start: 221
                                    end: 225
                                  ty:
                                    Primitive: I32
                            span:
                              start: 211
                              end: 225
                            ty: Unit
                          - kind:
                              Binary:
                                operator:
                                  kind: SubAssign
                                  span:
                                    start: 239
                                    end: 241
                                left:
                                  kind:
                                    Local:
                                      scope_index: 0
                                      local_index: 1
                                  span:
                                    start: 235
                                    end: 238
                                  ty:
                                    Primitive: I32
                                right:
                                  kind:
                                    Int: 1
                                  span:
                                    start: 242
                                    end: 243
                                  ty:
                                    Primitive: I32
                            span:
                              start: 235
                              end: 243
                            ty: Unit
                        result: ~
                    span:
                      start: 136
                      end: 250
                    ty:
                      Primitive: I32
              span:
                start: 131
                end: 250
              ty:
                Primitive: I32
        span:
          start: 46
          end: 252
        ty:
          Primitive: I32
  enums: []
  globals: []
  exports:
    - Function:
        func_index: 0
diagnostics: []
